wasm-runtime/polkadot/src/runtime/governance.rs

```diff
@@ -0,0 +1,361 @@
+//! the proposal then it is enacted. All items in storage concerning the proposal are reset.
+
+use runtime_support::prelude::*;
+use keyedvec::KeyedVec;
+use storable::{Storable, StorageVec, kill};
+use primitives::{AccountID, Hash, BlockNumber};
+use proposal::Proposal;
+use runtime::{staking, system, session};
+
+// TRANSACTION API
+
+/// Propose a sensitive action to be taken. Any action that is enactable by `Proposal` is valid.
+/// Proposal is by the `transactor` and will automatically count as an approval. Transactor must
+/// be a current validator. It is illegal to propose when there is already a proposal in effect.
+pub fn propose(validator: &AccountID, proposal: &Proposal) {
+	if Proposal::lookup(b"gov:pro").is_some() {
+		panic!("there may only be one proposal per era.");
+	}
+	proposal.store(b"gov:pro");
+	approve(validator, staking::current_era());
+}
+
+/// Approve the current era's proposal. Transactor must be a validator. This may not be done more
+/// than once for any validator in an era.
+pub fn approve(validator: &AccountID, era_index: BlockNumber) {
+	if era_index != staking::current_era() {
+		panic!("approval vote applied on non-current era.")
+	}
+	if Proposal::lookup(b"gov:pro").is_none() {
+		panic!("there must be a proposal in order to approve.");
+	}
+	if session::validators().into_iter().position(|v| &v == validator).is_none() {
+		panic!("transactor must be a validator to approve.");
+	}
+	let key = validator.to_keyed_vec(b"gov:app:");
+	if bool::lookup(&key).is_some() {
+		panic!("transactor may not approve a proposal twice in one era.");
+	}
+	true.store(&key);
+}
+
+/// Set the proportion of validators that must approve for a proposal to be enacted at the end of
+/// its era. The value, `ppm`, is measured as a fraction of 1000 rounded down to the nearest whole
+/// validator. `1000` would require the approval of all validators; `667` would require two-thirds
+/// (or there abouts) of validators.
+pub fn set_approval_ppm_required(ppm: u32) {
+	ppm.store(b"gov:apr");
+}
+
+// INSPECTION API
+
+/// The proportion of validators required for a propsal to be approved measured as the number out
+/// of 1000.
+pub fn approval_ppm_required() -> u32 {
+	Storable::lookup(b"gov:apr").unwrap_or(1000)
+}
+
+/// The number of concrete validator approvals required for a proposal to pass.
+pub fn approvals_required() -> u32 {
+	approval_ppm_required() * session::validator_count() as u32 / 1000
+}
+
+// PUBLIC API
+
+/// Current era is ending; we should finish up any proposals.
+pub fn end_of_an_era() {
+	// tally up votes for the current proposal, if any. enact if there are sufficient approvals.
+	if let Some(proposal) = Proposal::lookup(b"gov:pro") {
+		kill(b"gov:pro");
+		let approvals_required = approvals_required();
+		let approved = session::validators().into_iter()
+			.filter_map(|v| bool::take(&v.to_keyed_vec(b"gov:app:")))
```

Comment:
+1 for constans, `gov:apr` vs `gov:app:` is quite difficult to even distinguish visually. also what's `app` ro `apr`? why not full words?

Reply:
sure - it's a fairly big job though so i'd prefer to leave it for a later PR.